File: agentkit/prism_agentkit/provider.py
import os
from typing import Any

from coinbase_agentkit import ActionProvider, WalletProvider, create_action
from coinbase_agentkit.network import Network
from prismnetwork import DEFAULT_IMAGE, PrismAgent, PrismToolset
from prismnetwork.toolkit import UNSET
from pydantic import BaseModel, Field

DEFAULT_ESCROW = "0xfD4228eEEfC49e4b76A0CD40af9fdd546220B2FD"


class NoArgs(BaseModel):
    pass


class LeaseAndRunArgs(BaseModel):
    command: str = Field(description="Command to run on the rented GPU")
    duration_seconds: int = Field(600, description="How long to hold the lease, in seconds")
    min_vram_mib: int = Field(16000, description="Minimum GPU memory in MiB")
    image: str = Field(DEFAULT_IMAGE, description="Digest-pinned container image to boot")
    max_usdg: float | None = Field(
        None,
        description=(
            "Lower the operator's per-call spending cap for this one lease. It cannot raise it; "
            "left out, the operator's cap applies."
        ),
    )
    min_trust_class: str = Field(
        "open",
        description=(
            "Refuse suppliers below this trust class: open, isolated, attested or confidential. "
            "On an open supplier the host operator can read anything the workload touches."
        ),
    )


class RunArgs(BaseModel):
    lease_id: int = Field(description="A lease id returned by lease_and_run")
    command: str = Field(description="Command to run")


class LeaseIdArgs(BaseModel):
    lease_id: int = Field(description="A lease id from this session")


class PrismActionProvider(ActionProvider[WalletProvider]):
    """Rent and run on real NVIDIA GPUs through Prism Network.

    The provider carries its own funded wallet (``PRISM_AGENT_KEY``) and settles
    onchain in USDG. It never touches the agent's primary wallet, so it works with
    any AgentKit wallet provider.

    Spending goes through the same ledger the MCP server and the Python SDK write.
    ``max_usdg`` is the model's request rather than the limit: ``PRISM_MAX_USDG``
    bounds every lease, ``PRISM_DAILY_BUDGET_USDG`` bounds the day, and one wallet
    has one daily ceiling however many clients are holding it. ``budget`` takes a
    :class:`~prismnetwork.Budget` or :class:`~prismnetwork.SpendLedger` for a caller
    that wants its own limits.
    """

    def __init__(self, agent: PrismAgent | None = None, budget=UNSET):
        super().__init__("prism", [])
        if agent is None:
            key = os.environ.get("PRISM_AGENT_KEY")
            if not key:
                raise ValueError("PRISM_AGENT_KEY is required (or pass agent=)")
            agent = PrismAgent(key, os.environ.get("PRISM_ESCROW", DEFAULT_ESCROW))
        self.agent = agent
        self.tools = PrismToolset(agent=agent, budget=budget)

    @create_action(
        name="budget",
        description=(
            "Show the spending limits in force and what this wallet has already spent in the last "
            "24 hours. Worth checking before a long job; a lease refused for budget reports the "
            "same numbers."
        ),
        schema=NoArgs,
    )
    def budget(self, args: dict[str, Any]) -> str:
        return self.tools.budget_status()

    @create_action(
        name="wallet",
        description="Show the Prism agent's wallet address and its USDG and ETH balances.",
        schema=NoArgs,
    )
    def wallet(self, args: dict[str, Any]) -> str:
        return self.tools.wallet()

    @create_action(
        name="list_gpus",
        description=(
            "List GPUs available to rent right now, with model, VRAM, price per hour, and trust "
            "class. On an 'open' supplier the host operator can read anything the workload touches."
        ),
        schema=NoArgs,
    )
    def list_gpus(self, args: dict[str, Any]) -> str:
        return self.tools.list_gpus()

    @create_action(
        name="lease_and_run",
        description=(
            "Rent a GPU, run one command on it, and return the output. Pays onchain in USDG within "
            "the operator's per-call and daily spending caps. Blocks while the machine provisions "
            "(usually 1-4 minutes)."
        ),
        schema=LeaseAndRunArgs,
    )
    def lease_and_run(self, args: dict[str, Any]) -> str:
        p = LeaseAndRunArgs(**args)
        return self.tools.lease_and_run(
            command=p.command,
            duration_seconds=p.duration_seconds,
            min_vram_mib=p.min_vram_mib,
            image=p.image,
            max_usdg=p.max_usdg,
            min_trust_class=p.min_trust_class,
        )

    @create_action(
        name="run",
        description="Run another command on a GPU already leased in this session.",
        schema=RunArgs,
    )
    def run(self, args: dict[str, Any]) -> str:
        p = RunArgs(**args)
        return self.tools.run(p.lease_id, p.command)

    @create_action(
        name="end_lease",
        description="Release a leased GPU. Billing stops at the release; settlement charges the seconds it was open and returns the rest of the deposit.",
        schema=LeaseIdArgs,
    )
    def end_lease(self, args: dict[str, Any]) -> str:
        p = LeaseIdArgs(**args)
        return self.tools.end_lease(p.lease_id)

    def supports_network(self, network: Network) -> bool:
        return True


def prism_action_provider(agent: PrismAgent | None = None, budget=UNSET) -> PrismActionProvider:
    return PrismActionProvider(agent, budget)

File: agentkit-ts/provider.mjs
import { ActionProvider } from "@coinbase/agentkit";
import { DEFAULT_IMAGE, PrismAgent, TRUST_CLASSES } from "@prismnetwork/agent-sdk";
import { z } from "zod";

const MICROS_PER_USDG = 1_000_000;
const SECONDS_PER_HOUR = 3_600;

const noArgs = z.object({});

const leaseAndRunArgs = z.object({
  command: z.string().min(1).describe("Shell command to run on the rented GPU"),
  durationSeconds: z.number().int().min(60).max(21_600).default(600).describe("How long to hold the lease"),
  minVramMib: z.number().int().min(1).default(16_000).describe("Minimum GPU memory in MiB"),
  image: z.string().default(DEFAULT_IMAGE).describe("Digest-pinned container image to boot"),
  maxUsdg: z.number().positive().default(1).describe("Hard cap on the USDG this lease may cost"),
  minTrustClass: z
    .enum(TRUST_CLASSES)
    .default("open")
    .describe(
      "Refuse suppliers below this trust class. On an 'open' supplier the host operator can read anything the workload touches.",
    ),
});

const runArgs = z.object({
  leaseId: z.number().int().describe("Lease returned by lease_and_run"),
  command: z.string().min(1).describe("Shell command to run"),
});

const endLeaseArgs = z.object({
  leaseId: z.number().int().describe("Lease to release"),
});

function usdg(micros) {
  return `${(Number(micros) / MICROS_PER_USDG).toFixed(6)} USDG`;
}

function perHour(ratePerSecond) {
  return `$${((Number(ratePerSecond) * SECONDS_PER_HOUR) / MICROS_PER_USDG).toFixed(2)}/hr`;
}

function fromEnvironment() {
  const privateKey = process.env.PRISM_AGENT_KEY;
  const escrow = process.env.PRISM_ESCROW;
  if (!privateKey || !escrow) {
    throw new Error("PRISM_AGENT_KEY and PRISM_ESCROW are required, or pass a PrismAgent");
  }
  return new PrismAgent({ privateKey, escrow });
}

/**
 * Prism actions for AgentKit. Prism pays with its own wallet on Robinhood
 * Chain, so the wallet provider AgentKit is configured with is untouched.
 */
export class PrismActionProvider extends ActionProvider {
  #agent;
  #leases = new Map();

  constructor(agent) {
    super("prism", []);
    this.#agent = agent ?? fromEnvironment();
  }

  supportsNetwork() {
    return true;
  }

  getActions() {
    return [
      {
        name: "prism_wallet",
        description: "Report the Prism wallet address and its USDG and gas balances.",
        schema: noArgs,
        invoke: async () => {
          const { address, usdg: balance, eth } = await this.#agent.balances();
          return `${address}\n${usdg(balance)}\n${(Number(eth) / 1e18).toFixed(6)} ETH for gas`;
        },
      },
      {
        name: "prism_list_gpus",
        description:
          "List GPUs available to rent right now, with model, memory, price per hour, and trust class. On an 'open' supplier the host operator can read anything the workload touches.",
        schema: noArgs,
        invoke: async () => {
          const offers = await this.#agent.offers();
          if (!offers.length) return "No GPUs are available right now.";
          return offers
            .map(
              (offer) =>
                `${offer.gpu.model} · ${offer.gpu.vram_mib} MiB · ${perHour(offer.rate_per_second)} · ${offer.trust_class ?? "open"}`,
            )
            .join("\n");
        },
      },
      {
        name: "prism_lease_and_run",
        description:
          "Rent a GPU, run one command on it over SSH, and keep the lease open for follow-up commands. Funds an onchain escrow.",
        schema: leaseAndRunArgs,
        invoke: async (args) => {
          const lease = await this.#agent.lease({
            image: args.image,
            durationSeconds: args.durationSeconds,
            minVramMib: args.minVramMib,
            maxDeposit: Math.round(args.maxUsdg * MICROS_PER_USDG),
            minTrustClass: args.minTrustClass,
          });
          this.#leases.set(lease.leaseId, lease);
          const result = await this.#agent.run(lease, args.command);
          return [
            `lease ${lease.leaseId} funded onchain: ${lease.fundingHash}`,
            `exit ${result.code}`,
            result.stdout.trim(),
            result.stderr.trim(),
          ]
            .filter(Boolean)
            .join("\n");
        },
      },
      {
        name: "prism_run",
        description: "Run another command on a GPU already leased in this session.",
        schema: runArgs,
        invoke: async (args) => {
          const lease = this.#leases.get(args.leaseId);
          if (!lease) return `lease ${args.leaseId} is not open in this session`;
          const result = await this.#agent.run(lease, args.command);
          return [`exit ${result.code}`, result.stdout.trim(), result.stderr.trim()].filter(Boolean).join("\n");
        },
      },
      {
        name: "prism_end_lease",
        description:
          "Release a lease. Billing stops at the release; settlement charges the seconds it was open and returns the rest of the deposit.",
        schema: endLeaseArgs,
        invoke: async (args) => {
          const lease = this.#leases.get(args.leaseId);
          if (!lease) return `lease ${args.leaseId} is not open in this session`;
          this.#leases.delete(args.leaseId);
          const out = await this.#agent.endLease(lease);
          if (out?.release === "failed") {
            return `lease ${args.leaseId} could not be released: ${out.error}. Its access key is gone but the meter may still be running; check receipts for the settled charge.`;
          }
          return `released lease ${args.leaseId}; billing stopped here and the unused deposit returns after settlement`;
        },
      },
    ];
  }
}

export function prismActionProvider(agent) {
  return new PrismActionProvider(agent);
}
